File: java/retta.java
import java.util.Vector;
import static java.lang.Math.*;

class retta extends inputitem {
	private razionale a, b , c ;

	public retta(razionale x,razionale y, razionale l){
		a = x;
		b = y;
		c = l;
	}

	public retta(){
		this(new razionale(),new razionale(),new razionale());	
	}

	public retta(double n, double m,double o){
		this(new razionale(n),new razionale(m),new razionale(o));
	}

	public razionale getA(){
		return a;
	}

	public razionale getB(){
		return b;
	}

	public razionale getC(){
		return c;
	}

	public String toString(){
		String ret = new String();

		ret = ret + "( ";
		
		if( a.getnum() != 0 ) ret = ret + a + " x " ;

		if( b.getnum() != 0 ) {
			if( b.sign() == true ) ret = ret + " + ";

			ret = ret + b  + " y " ;
		}

		if( c.getnum() != 0 ){
			if( c.sign() == true ) ret = ret + " + ";

		    ret = ret + c ;
		}

		ret = ret + " )";


		return ret;
	}

	public double distancePuntoRetta(punto p){

		return  ( (abs( (a.multiply(p.getx())).converti() + (b.multiply(p.gety())).converti() + c.converti() ) )  /(sqrt(pow(a.converti(),2) + pow(b.converti(),2))));
	}

	public double distanceRettaRetta(retta r){
		punto x = new punto();
		razionale neg = new razionale(-1,1);
		razionale zero = new razionale();

		if(r.a.getnum() == 0){
	        x = new punto(zero,neg.multiply(r.c.multiply(r.b.inverso())));
	    }
	    else if(r.b.getnum() == 0 ){
	        x = new punto(((r.c.multiply(r.a.inverso())).multiply(neg)),zero);
	    }
	    else{

	        x = new punto(zero,(neg.multiply(r.c.multiply(r.b.inverso()))));
	    }
	    return distancePuntoRetta(x);
	}


	public boolean isParallels(retta r2 ){
		razionale neg = new razionale(-1,1);
		razionale zero = new razionale();

	    double coeffA = neg.multiply((this.a.multiply(this.b.inverso()))).converti();
	    double coeffB = neg.multiply((r2.a.multiply(r2.b.inverso()))).converti();
	    if(coeffA == coeffB) return true;
	    else return false;
	}

	public boolean isPerpendicolari( retta r2){
		razionale neg = new razionale(-1,1);
	    if( neg.multiply((this.a.multiply(this.b.inverso()))).equals(neg.multiply(neg.multiply((r2.a.multiply(r2.b.inverso())).inverso()))) )
	        return true;
	    else return false;
	}

	//retta perpendicolare a un altra e passante per un punto
	public retta rettaperpendicolare(punto p ){
		razionale neg = new razionale(-1,1);
		razionale zero = new razionale();

	    if( b.getnum() != 0 && a.getnum() != 0 ){
	        //trovo il coefficente angolare
	        razionale m = new razionale(a.multiply(neg),b);
	        //trovo l'antireciproco del coefficente angolare
	        razionale new_m =  (m.inverso()).multiply(neg);
	        //trovo c => y1 + m*x1
	        razionale c = new_m.multiply(p.getx());
	        c = new razionale( (c.multiply(neg)).converti() + (p.gety()).converti());
	        return new retta(new_m,neg,c);
	    }else{
	        if( a.equals(0) ){
	           return new retta( neg,zero,p.getx() );
	        }
	        else return new retta( zero,neg,p.gety() );
	    }
	}

	public retta rettaparallela(punto p){
		razionale neg = new razionale(-1,1);
		razionale zero = new razionale();

		if( b.getnum() != 0 && a.getnum() != 0){ //disuguaglianza double ==> primitivi
	        //trovo il coefficente angolare
	        razionale m = new razionale( a.multiply(neg),b );
	        //trovo c => y1 + m*x1
	        razionale c = m.multiply(p.getx()) ;

	        c = (c.multiply(neg)).add(p.gety());

	        return new retta(m,neg,c);
	    }
	    else{
	        if(getA().getnum() == 0){
	            return new retta(zero,neg,p.gety());
	        }
	        else return new retta(neg,zero,p.getx());
	    }
	}


	//dato una x ritorna il punto rispettivo 
	public punto printCoord_x(razionale x) throws eccezioni{
		razionale neg = new razionale(-1,1);

		razionale y = new razionale(1,1);

		if(getB().getnum() == 0 && getA().getnum() == 0){
	       throw new implicit();
	    }
	    else if(getB().getnum() == 0){
	        x = new razionale ( getC().multiply(neg) , getA() );

	    }
	    else{
	        y = new razionale( (getA().multiply(neg)).multiply(x).add(getC().multiply(neg)) ,getB());
	    }

	    return new punto(x,y);
	}

	//mi serve per disegnare... la facciooo ?
	/*public Vector<punto> print_rect() {

	}*/

	public Vector<punto> intersectretta(retta r1){
		Vector<punto> p = new Vector<punto>();
		razionale neg = new razionale(-1,1);

		if(isParallels(r1) == true) return p;

		razionale det = new razionale ( getA().multiply(r1.getB()).add( neg.multiply(r1.getA().multiply(getB()) ) ));

		if(det.getnum() != 0){

			razionale detx = new razionale(((getC().multiply(neg)).multiply(r1.getB())).add(neg.multiply((r1.getC().multiply(neg)).multiply(getB()))),det) ;

	        razionale dety = new razionale(((getA().multiply(neg)).multiply(r1.getC())).add(neg.multiply((r1.getA().multiply(neg)).multiply(getC()))),det) ;

	        p.add(new punto(detx,dety));	        
		}
		else{
			razionale x = new razionale( ((getC().multiply((r1.getB()).multiply(getB().inverso()))).add(neg.multiply(r1.getC()))),(r1.getA()).add(neg.multiply(getA().multiply((r1.getB()).multiply(getB().inverso())))));
	        razionale y = new razionale((getA().multiply(x)).multiply(neg.multiply(getB())).add(getC().multiply((neg.multiply(getB())).inverso())));
	        p.add(new punto(x,y));
		}
		return p;
	}

	public double distance(inputitem i) throws eccezioni{
		if(i instanceof retta){
			if((this.intersect(i)).size() == 1) return 0;
			retta r = (retta)i;
			return distanceRettaRetta(r);
		}
		else if(i instanceof punto){
			punto p = (punto)i;
			return distancePuntoRetta(p);
		}
		else{
			return i.distance(this);
		}
	}

	public Vector<punto> intersect(inputitem i){
		Vector<punto> v = new Vector<punto>();
		
		if(i instanceof retta){
			retta r = (retta)i;
			return intersectretta(r);
		}
		else if(i instanceof punto){
			punto p = (punto)i ;
			if(distancePuntoRetta(p) == 0) {
				v.add(p);
			}
		}
		else{
			return i.intersect(this);
		}

		return v;//i.intersect(this);
	}

	public void pars_rect(String retta) throws eccezioni
	{
		retta = retta.replace(" ","");

		//inserisco = a prescindere
	    retta = retta + '=';

	    String s = new String();
	    retta r;

	    // raz = verifica se è una frazione .
	    // incx = incy = serve per verificare se trova due volte la
	    // la x o la y .
	    boolean raz = false , incx = false , incy = false , doub = false , segno = false ;

	    //n = num , d = den , x = coeff x , y = coeff y , tn = termine noto
	    double n=0,d=0,x=0,y=0,tn=0;

	    int sign = 1;
	    double virgola = 1 , n1;

	    String rect[] = new String[retta.length()];
	    rect = retta.split("");

	    for(int i = 0 ; i < rect.length; i++){

	        //vado a verificare ci sia solo una x e una y altrimenti non e' nella forma prevista
	        if(rect[i].equals("y") && incy == false) incy = true ;
	        else if(rect[i].equals("y") && incy == true) throw new implicit();

	        if(rect[i].equals("x") && incx == false) incx = true ;
	        else if(rect[i].equals("x") && incx == true) throw new implicit();

	        if( !rect[i].equals("*") && !rect[i].equals("=") )
	        {
	            if( rect[i].equals("-") || rect[i].equals("+") ){
	                //verifico se il termine noto è a primo o secondo membro

	                if(s.length() > 0){
	                    n1 = Integer.parseInt(s);
	                    tn = n1 / virgola;

	                    virgola = 1;
	                    doub = false;

	                    if( ( !(b.getnum() == 0) || (a.getnum() == 0) ) && segno == false )  throw new input_error();

	                    c = new razionale(sign*tn,1);
	                    sign = 1;
	                    segno = false;
	                    s = "";
	                }
	                if( rect[i].equals("-") ) sign = -1;

	                segno = true;

	            }
	            else if( rect[i].equals("/") ){

	                n1 = Integer.parseInt(s);
	                n = n1 / virgola;

	                virgola = 1;
	                doub = false;

	                s = "";
	                raz = true;
	            }
	            else {
	                //caso in cui ci sia un x o y al denominatore
	                if( ( rect[i].equals("x") || rect[i].equals("y") ) && raz == true )
	                     throw new input_error();

	                if( rect[i].equals("x") || rect[i].equals("y") ){
	                    if( s.length() == 0) s="1";

	                    n1 = Integer.parseInt(s);
	                    x = n1 / virgola;

	                    virgola = 1;
	                    doub = false;
	                    if((b.getnum() != 0 || c.getnum() != 0) && segno == false) throw new input_error();

	                    if(rect[i].equals("x"))
	                    	a = new razionale(sign*x,1);
	                    else 
	                    	b = new razionale(sign*x,1);

	                    sign = 1;
	                    segno = false;
	                    s = "";
	                }
	                else{
	                    if(raz == true){
	                        //vuol dire che sono a denominatore
	                        while( !rect[i].equals("x") && !rect[i].equals("y") && !rect[i].equals("=") && !rect[i].equals("+") && !rect[i].equals("-") )
	                        {
	                            if(doub == true) virgola *= 10 ;
	                            if( rect[i].equals(".") ){
	                                doub = true;
	                            }
	                            else s = s + rect[i];
	                            i++;
	                        }

	                        n1 = Integer.parseInt(s);
	                        d = n1 / virgola;

	                        virgola = 1;
	                        doub = false;

	                        if( rect[i].equals("x") ){
	                            if((b.getnum() != 0 || c.getnum() != 0) && segno == false )  throw new input_error();
	                            a = new razionale(sign*n,d);
	                            segno = false;
	                            sign = 1;
	                        }else if( rect[i].equals("y") ){
	                            if((a.getnum() != 0 || c.getnum() != 0) && segno == false ) throw new input_error(); 
	                            b = new razionale(sign*n,d);
	                            sign = 1;
	                            segno = false;
	                        }else {
	                            if( (a.getnum() != 0 || b.getnum() != 0) && segno == false )  throw new input_error();
	                            c = new razionale(sign*n,d);
	                            sign = 1;
	                            segno = false;
	                        }

	                        //decremento cosi posso ripartire il ciclo dal segno
	                        if(rect[i].equals("-") || rect[i].equals("+")) --i;
	                        n=0;d=1;
	                        s = "";
	                        raz = false;
	                    }
	                    else{


	                        //verifico se rect[i] == numero
	                        if( rect[i].equals('.') ){
	                            doub = true;
	                        }
	                        else {
	                        	try{
			                		Integer.parseInt(rect[i]);
				                }
				                catch(java.lang.NumberFormatException nfe){
				                	throw new not_numeric();
				                }
				                if(doub == true) virgola *= 10 ;
                          		s = s + rect[i];
		                    }
	                        //inserito un input errato: carattere non riconosciuto
	                    }
	                }
	            }
	        }
	        else if( rect[i].equals('=') || i == rect.length - 1 ){
	            //termine noto ==> tn
	             if(s.length() > 0){
	                 n1 = Integer.parseInt(s);
	                 tn = n1 / virgola;

	                 virgola = 1;
	                 doub = false;

	                 if(( !(b.getnum() == 0) || !(a.getnum() == 0) ) && segno == false) throw new input_error();

	                 c = new razionale(sign*tn,1);
	                 segno = false;

	                 s = "";
	              }

	              sign = 1;
	              break;

	        }else {
	            s = "";
	            n=0;d=0;
	        }
	    }
	    if( a.getnum() == 0 && b.getnum()== 0) {
	        throw new input_error();
	    }

	}


}

File: c++/model/retta.cpp
#include "retta.h"
#include <iostream>
#include <locale>
#include <sstream>
using std::vector;

razionale retta::GetA() const { return a; }

razionale retta::GetB() const { return b; }

razionale retta::GetC() const { return c; }

double retta::distancePuntoRetta(const punto& p) const{
    return ((abs(a * p.getX() + b * p.getY() + c))/(sqrt(pow(a,2)+pow(b,2))));
}

double retta::distanceRettaRetta(retta& r2) const {
    //trovo il punto con x = 0 di r2:
    if(r2.GetA() == 0){
        punto x (0,(-1)*(r2.GetC()/r2.GetB()));
        return distancePuntoRetta(x);
    }
    else if(r2.GetB() == 0){
        punto x ((r2.GetC()/r2.GetA())*(razionale(-1,1)),0);
        return distancePuntoRetta(x);
    }
    else{
        punto x (0,(-1)*(r2.GetA()/r2.GetB())*(0) - r2.GetC()/r2.GetB());
        return distancePuntoRetta(x);
    }

}

double retta::distance(inputitem * i) const {
    if((i->intersect(const_cast<retta*>(this))).size() == 1) return 0;


    if(typeid(*i) == typeid(retta)){
        return distanceRettaRetta(*(dynamic_cast<retta*>(i)));
    }
    else if( typeid(punto) == typeid(*i) ){
        return distancePuntoRetta(*(dynamic_cast<punto*>(i)));
    }
    else return i->distance(const_cast<retta*>(this));
}

retta retta::rettaFromTwoPoints(const punto& p1,const punto& p2){
    if(p1.getX() == p2.getX() && p1.getY() != p2.getY()){
        return retta(-1,0,p1.getX());
    }
    else if(p1.getX() != p2.getX() && p1.getY() == p2.getY()){
        return retta(0,-1,p1.getY());
    }
    else{
        razionale d1(p2.getX() - p1.getX());
        razionale d2(p2.getY() - p1.getY());
        razionale a = d2;
        razionale b = d1;
        razionale c = (d2*p1.getX()*(razionale(-1,1))) + (d1*p1.getY());
        return retta(a,b*(razionale(-1,1)),c);
    }
}

bool retta::isParallels(retta& r1 ) const{
    razionale coeffA = razionale((razionale(-1,1))*(r1.GetA()/r1.GetB()));
    razionale coeffB = razionale((razionale(-1,1))*(GetA()/GetB()));

    if( (r1.GetA() == 0 && GetB() == 0) || (r1.GetB() == 0 && GetA() == 0) ){
        return false;
    }

    return coeffA == coeffB;
}

bool retta::isPerpendicolari(retta& r1 ) const{
    if((-1)*(r1.GetA()/r1.GetB()) == -1/((-1)*(GetA()/GetB())))
        return true;
    else return false;
}

retta retta::RettaPerpendicolare( punto& p ) const{
    //m è già il nuovo coefficente angolare
    if(GetB() != 0 && GetA() != 0){
        //trovo il coefficente angolare
        razionale m(GetA()*(razionale(-1,1)),GetB());
        //trovo l'antireciproco del coefficente angolare
        razionale new_m = m.inverso()*(razionale(-1,1));
        //trovo c => y1 + m*x1
        razionale c(new_m*p.getX());
        c=(c*(razionale(-1,1)))+p.getY();
        return retta(new_m,-1,c);
    }else{
        if(GetA() == 0){
           return retta(-1,0,p.getX());
        }
        else return retta(0,-1,p.getY());
    }
}

retta retta::RettaParallela(punto& p ) const{
    if(GetB() != 0 && GetA() != 0){
        //trovo il coefficente angolare
        razionale m(GetA()*(razionale(-1,1)),GetB());
        //trovo c => y1 + m*x1
        razionale c(m*p.getX());
        c=(c*(razionale(-1,1)))+p.getY();
        return retta(m,-1,c);
    }
    else{
        if(GetA() == 0){
            return retta(0,-1,p.getY());
        }
        else return retta(-1,0,p.getX());
    }
}

vector<punto> retta::Intersect_rette(const retta& r2) const {
    razionale Det = (GetA()*r2.GetB()) - (r2.GetA()*GetB());
    vector<punto> p;

    if(r2.isParallels(const_cast<retta&>(*this))) return p;

    

    if(Det != 0){
        razionale DetX = razionale( GetC()*(razionale(-1,1))*r2.GetB() + razionale(-1,1)*(r2.GetC()*(razionale(-1,1))*GetB()) , Det);
        razionale DetY = razionale(GetA()*(razionale(-1,1))*r2.GetC() + razionale(-1,1)*(r2.GetA()*(razionale(-1,1))*GetC() ) , Det);
        p.push_back(punto(DetX,DetY));
        return p;
    }
    else
    {
        razionale x((((GetC()*r2.GetB())/GetB()) - r2.GetC()),((r2.GetA()-(GetA()*r2.GetB())/GetB())));
        razionale y = razionale((GetA()*x)/(razionale(-1,1)*(GetB())) + GetC()/(razionale(-1,1)*(GetB())));
        p.push_back(punto(x,y));
        return p;
    }
}

//ritorna un vector di razionali: indice dispari = x , indice pari = y
punto retta::printCoord_x(razionale x) const{
    if(GetB() == 0 && GetA() == 0){
        throw not_implicit();
    }
    else if(GetB() == 0){
        razionale x( GetC()*(razionale(-1,1)),GetA());
        return punto(x,razionale(1,1));
    }
    else{
        razionale y(((GetA()*(razionale(-1,1)))*x + GetC()*(razionale(-1,1))),GetB());
        return punto(x,y);
    }
}

//per ora distanza standard tra i punti : 1 cm (modificabile ??)
vector<punto> retta::print_rect(const razionale& r1 ,const razionale& r2){
    vector<punto> pt;
    if(GetB() == 0){
        punto p1(razionale(-1,1)*razionale(GetC(),GetA()),r1);
        punto p2(razionale(-1,1)*razionale(GetC(),GetA()),r2);
        pt.push_back(p1);
        pt.push_back(p2);
    }
    else{
        pt.push_back(printCoord_x(r1));
        pt.push_back(printCoord_x(r2));
    }
    return pt;
}

vector<punto> retta::intersect( inputitem *i ) const {
    vector<punto> punt ;

    if( typeid(*i) == typeid(retta) ){
        retta * r = dynamic_cast<retta*>(i);
        return r->Intersect_rette(*this);

    }else if( typeid(*i) == typeid(punto) ){
        punto * p = dynamic_cast<punto*>(i);
        if(distancePuntoRetta(*p) == 0 ){
            punt.push_back(*p);
        }
        return punt;
    }

    //i = poligono
    return i->intersect(const_cast<retta*>(this));
}

QString retta::toString()const{
    std::string st;

    if(a.GetNum() != 0) st = st + a.tostring() + "x ";
    if(b.GetNum() != 0) st = st + b.tostring() + "y ";
    if(c.GetNum() != 0) st = st + c.tostring();

    QString qstr = QString::fromStdString(st);

    return qstr;
}

//-----------------------OVERLOAD OPERATORI---------------------------------------------------

ostream& operator<<(ostream& buffer, const retta& r){
    buffer<<"(";
    if(r.a != 0)
        buffer<<r.a<<"x";

    if(r.b >= 0) buffer<<"+";
    if(r.b != 0){
        buffer<<r.b<<"y";
    }

    if(r.c >= 0) buffer<<"+";
    if(r.c != 0){
        buffer<<r.c;
    }

    buffer<<" =0 )";
    return buffer;
}


void retta::pars_rect(string rect)
{
    unsigned int len = rect.length();
    bool trovato = false ;
    for (unsigned int var = 0; var < len; ++var) {
        if(rect[var] == ' '){
            rect.erase(rect.begin()+var);
            var--;
            len--;
        }
        if(rect[var] == '=') trovato = true;
    }

    if( !trovato ) rect = rect + '=';

    std::string s;
    retta r;

    // raz = verifica se è una frazione .
    // incx = incy = serve per verificare se trova due volte la
    // la x o la y .
    bool raz = false , incx = false , incy = false , doub = false , segno = false ;

    //n = num , d = den , x = coeff x , y = coeff y , tn = termine noto
    double n=0,d=0,x=0,y=0,tn=0;

    int sign = 1;
    double virgola = 1 , n1;

    for(unsigned int i = 0 ; i < rect.length() ; i++){
        //vado a verificare ci sia solo una x e una y altrimenti non e' nella forma prevista
        if(rect[i] == 'y' && !incy) incy = true ;
        else if(rect[i] == 'y' && incy) throw not_implicit();

        if(rect[i] == 'x' && !incx) incx = true ;
        else if(rect[i] == 'x' && incx) throw not_implicit();

        if(rect[i] != '*' && rect[i] != '=')
        {
            if(rect[i] == '-' || rect[i] == '+'){
                //verifico se il termine noto è a primo o secondo membro

                if(s.length() > 0){
                    n1 = std::stoi( s );
                    tn = n1 / virgola;

                    virgola = 1;
                    doub = false;

                    if((b.GetNum() != 0 || a.GetNum() != 0) && !segno) throw input_error();
                    c = razionale(sign*tn,1);
                    sign = 1;
                    segno = false;
                    s.erase(s.begin(),s.end());
                }
                if(rect[i] == '-' && sign == 1) sign = -1;
                else if(rect[i] == '-' && sign == -1) sign = 1;

                segno = true;


            }
            else if(rect[i] == '/'){

                n1 = std::stoi( s );
                n = n1 / virgola;

                virgola = 1;
                doub = false;

                s.erase(s.begin(),s.end());
                raz = true;
            }
            else {
                //caso in cui ci sia un x o y al denominatore
                if( (rect[i] == 'x' || rect[i] == 'y') && raz == true )
                    throw input_error("x/y non può essere a denominatore");

                if(rect[i] == 'x'){
                    if(s.length() == 0) s='1';

                    n1 = std::stoi( s );
                    x = n1 / virgola;

                    virgola = 1;
                    doub = false;
                    if((b.GetNum() != 0 || c.GetNum() != 0) && !segno) throw input_error();

                    a = razionale(sign*x,1);
                    sign = 1;
                    segno = false;
                    s.erase(s.begin(),s.end());
                }else if(rect[i] == 'y'){
                    if(s.length() == 0) s='1';

                    n1 = std::stoi( s );
                    y = n1 / virgola;

                    virgola = 1;
                    doub = false;

                    if((a.GetNum() != 0 || c.GetNum() != 0) && !segno) throw input_error();
                    b = razionale(sign*y,1);
                    sign = 1;
                    segno = false;
                    s.erase(s.begin(),s.end());
                }
                else{
                    if(raz == true){
                        //vuol dire che sono a denominatore
                        while(rect[i] != 'x' && rect[i] != 'y' && rect[i] != '=' && rect[i] != '+' && rect[i] != '-')
                        {
                            if(doub) virgola *= 10 ;
                            if(rect[i] == '.'){
                                doub = true;
                            }
                            else s = s+rect[i];
                            i++;
                        }

                        n1 = std::stoi( s );
                        d = n1 / virgola;

                        virgola = 1;
                        doub = false;

                        if(rect[i] == 'x'){
                            if((b.GetNum() != 0 || c.GetNum() != 0) && !segno) throw input_error();
                            a = razionale(sign*n,d);
                            segno = false;
                            sign = 1;
                        }else if(rect[i] == 'y'){
                            if((a.GetNum() != 0 || c.GetNum() != 0) && !segno) throw input_error();
                            b = razionale(sign*n,d);
                            sign = 1;
                            segno = false;
                        }else {
                            if((a.GetNum() != 0 || b.GetNum() != 0) && !segno) throw input_error();
                            c = razionale(sign*n,d);
                            sign = 1;
                            segno = false;
                        }

                        //decremento cosi posso ripartire il ciclo dal segno
                        if(rect[i] == '-' || rect[i] == '+') --i;
                        n=0;d=1;
                        s.erase(s.begin(),s.end());
                        raz = false;
                    }
                    else{

                        std::locale loc;
                        //verifico se rect[i] == numero
                        if(rect[i] == '.'){
                            doub = true;
                        }
                        else if(isdigit(rect[i],loc))
                        {
                          if(doub) virgola *= 10 ;
                          s = s+rect[i];
                        }
                        else throw input_error("Carattere non riconosciuto.");
                        //inserito un input errato: carattere non riconosciuto
                    }
                }
            }
        }
        else if(rect[i] == '=' || rect[i+1] == '='){
            //termine noto ==> tn
             if(s.length() > 0){
                 n1 = std::stoi( s );
                 tn = n1 / virgola;

                 virgola = 1;
                 doub = false;

                 if((b.GetNum() != 0 || a.GetNum() != 0) && !segno) throw input_error();

                 c = razionale(sign*tn,1);
                 segno = false;

                 s.erase(s.begin(),s.end());
              }
              sign = 1;
              break;
        }else {
            s.erase(s.begin(),s.end());
            n=0;d=0;
        }
    }
    if(a.GetNum() == 0 && b.GetNum() == 0){
        throw input_error("Mancano le incognite x y .");
    }
    else{
        throw 1;
    }


}



